### PR TITLE
Input: add counter same as text area #185

### PR DESCRIPTION
**Pull request by mozesoron**
### What changed
added new counter same as text area
...

### Why it changed
#185
...

---

- [ ] Change is tested
- [ ] Change is documented
- [ ] Build is green
- [ ] Change of UX is approved by [wuwa](https://github.com/wuwa) or [milkyfruit](https://github.com/milkyfruit)

### CODE REVIEW

File: src/Input/Input.driver.js [hunk 1]
```diff
@@ -76,7 +77,9 @@ const inputDriverFactory = ({element, wrapper, component}) => {
     setProps: props => {
       const ClonedWithProps = React.cloneElement(component, Object.assign({}, component.props, props), ...(component.props.children || []));
       ReactDOM.render(<div ref={r => element = r}>{ClonedWithProps}</div>, wrapper);
-    }
+    },
+    getHasCounter: () => !!element.querySelectorAll(counterSelector).length,
```
Comment (lbelinsk) [inline src/Input/Input.driver.js]: Can you please just rename it to `hasCounter`. It is a little weird like that :)

File: src/Input/Input.spec.js [hunk 2]
```diff
@@ -602,6 +602,18 @@ describe('Input', () => {
...
     });
   });
+
+  describe('counter', () => {
+    it('should show correct value when hasCounter and maxLength present', () => {
+      const driver = createDriver(<Input hasCounter maxLength={30} value={'abc'}/>);
+      expect(driver.getCounterValue()).toEqual('3/30');
+    });
+
+    it('should not show counter when hasCounter is not present', () => {
```
Comment (lbelinsk) [inline src/Input/Input.spec.js]: I would also test when only maxLength is not present, and hasCounter is present (division by zero that you handled in the code :))

File: src/Input/Input.js [hunk 3]
```diff
@@ -393,6 +396,8 @@ Input.propTypes = {
   tooltipPlacement: PropTypes.string,
   type: PropTypes.string,
   unit: PropTypes.string,
+  /** When true a letters counter will appear */
+  hasCounter: PropTypes.bool,
```
Comment (lbelinsk) [inline src/Input/Input.js]: I think a more react prop fitting name can be nicer here. I just can't think of one. If you find one, great, if not, never mind :)

File: src/Input/Input.spec.js [hunk 4]
```diff
@@ -602,6 +602,18 @@ describe('Input', () => {
...
+      const driver = createDriver(<Input/>);
+      expect(driver.getHasCounter()).toBeFalsy();
+    });
+  });
```
Comment (lbelinsk) [inline src/Input/Input.spec.js]: I would also test the case where we start with defaultValue, to make sure the counter is correct

File: src/Input/Input.js [hunk 5]
```diff
@@ -137,6 +139,7 @@ class Input extends Component {
     //https://github.com/wix/wix-style-react/issues/1691
     return (<div className={styles.inputWrapper}>
       {prefix && <div className={styles.prefixSuffixWrapper}><div className={styles.prefix}>{prefix}</div></div>}
+      {hasCounter && maxLength && <span className={styles.counter} data-hook="input-counter">{this.state.counter}/{maxLength}</span>}
```
Comment (lbelinsk) [inline src/Input/Input.js]: why do you need to keep the counter on the state?
{this.props.value.length}/{maxLength}

Is it due to performance reasons, not to calculate the length on each render?
I think it is a tiny operation, we can remove from the state. wdyt?

**Comment by lbelinsk** [comment]
@mozesoron Thanks for the pr.
I left some comments. Is there a reason why you didn't use the `prefix` prop for that?
You tagged some pr of breadcrumbs, but I didn't understand the context... What is this pr for? Did someone approve the design?

@milkyfruit Is this approved? This pr is adding a word counter field to the input as a prefix.
image

File: src/Input/Input.driver.js (hunk 1 above)
Comment (mozesoron) [reply src/Input/Input.driver.js]: this is the same name as in the textarea - I think it is better to use the same names for the same func. WDYT?

File: src/Input/Input.js (hunk 3 above)
Comment (mozesoron) [reply src/Input/Input.js]: same as in textarea

File: src/Input/Input.js (hunk 5 above)
Comment (mozesoron) [reply src/Input/Input.js]: so same as other comments - implemented same as in textarea. also the counter should be on the component state. it is internal and related only to the component

File: src/Input/Input.spec.js (hunk 4 above)
Comment (mozesoron) [reply src/Input/Input.spec.js]: I'v extended the test case

File: src/Input/Input.driver.js (hunk 1 above)
Comment (lbelinsk) [reply src/Input/Input.driver.js]: I guess so.

File: src/Input/Input.js (hunk 5 above)
Comment (lbelinsk) [reply src/Input/Input.js]: It is related to the component, sure, The user won't need to take care of it. but it is only data that you have on the prop, no need to keep on the state

File: src/Input/Input.scss [hunk 6]
```diff
@@ -15,6 +15,16 @@ $material-gap: 8px;
...
 }
 
+.counter{
+  position: absolute;
+  top: -30px;
+  right: 0;
+  font-size: 14px;
+  line-height: 1.3;
+  @include Placeholder($GR60);
+  color: #7a92a5;
```
Comment (lbelinsk) [inline src/Input/Input.scss]: Please use D40

File: src/Input/Input.spec.js (hunk 4 above)
Comment (lbelinsk) [reply src/Input/Input.spec.js]: but you need to add some expectation on the value as well...

**Comment by milkyfruit** [comment]
@mozesoron @lbelinsk this is not followed by the guidelines. The char counter is defined to be outside the input field:
image

Link to specs: https://zpl.io/2Egjola

**Comment by lbelinsk** [comment]
@mozesoron It means that this should be added to TextField, not Input component. 
https://wix-wix-style-react.surge.sh/?selectedKind=3.%20Inputs&selectedStory=3.1%20Text%20Field&full=0&addons=0&stories=1&panelRight=0

I am closing this pr just because it is not approved by the design.
I can leave the branch if it helps

If you would just copy the same implementation we have in InputArea to TextField I will not ask for changes or additional stuff. I will do them if needed, so please just add the way you see right

**Comment by shlomitc** [comment]
@argshook - I think the counter functionality should sit in the input, because the state of the characters length sits there and not in the form field component.

Having it in the field will create a complex API for the user for a simple use case. Also. Think about it like an error message. It could have been internally inside the input prefix and not separate.

**Comment by argshook** [comment]
@shlomitc no, disagree.

input supports prefix & suffix, they accept nodes and so user of `Input`
can decide what to display there, might as well be custom counter.

making a special case for a thing that does not exist in design system
is asking for trouble in the long run.

Field component API will be this:

```js
<Field
  maxLength={100}
  value={yourState.value}
  onChange={yourHandler}
  >
  <Input/>
</Field>
```

it's not complex at all.
I'll give it for review today.

if counter functionality sits in input, then it also has to sit in InputArea, then RichTextArea then possibly more.

unless this is urgent for @mozesoron , i don't see why rush with it, let's follow design system

**Comment by lbelinsk** [comment]
@argshook assigning you as a reviewer for this one to help you remember to close / add this fix in your generalized solution

**Comment by shlomitc** [comment]
@mozesoron - please see my comment [here](https://github.com/wix/wix-style-react/issues/1664#issuecomment-404683231)
if it fit your needs, let's close this PR

**Comment by shlomitc** [comment]
I see that you're good with the new component, so closing this PR. thanks!
